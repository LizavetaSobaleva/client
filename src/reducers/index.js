import { applyMiddleware, combineReducers } from "redux";
import { configureStore} from "@reduxjs/toolkit"
import { thunk } from "redux-thunk";
import userReducer from "./userReducer";
import fileReducer from "./fileReducer";
import uploadReducer from "./uploadReducer";
import appReducer from "./appReducer";
import { composeWithDevTools } from '@redux-devtools/extension';


const rootReducer = combineReducers({
    user: userReducer,
    files: fileReducer,
    upload: uploadReducer,
    app: appReducer
})

export const store = configureStore({reducer: rootReducer, devTools: composeWithDevTools(applyMiddleware(thunk))})