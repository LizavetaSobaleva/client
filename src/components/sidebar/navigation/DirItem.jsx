import React, { useState } from 'react'
import './navigation.less'
import { useDispatch} from 'react-redux'
import { pushToStack, setCurrentDir, clearStack } from '../../../reducers/fileReducer'
import DropdownButton from '../../UI/dropdownButton/DropdownButton'

const DirItem = ({file, title}) => {
  const dispatch = useDispatch()
  const [isOpen, setIsOpen] = useState(false)
  const hasChildren = file && file.childs && file.childs.length > 0;


  function openHandler() {
    const path = file.idPath.slice(0, -1)
    dispatch(clearStack())
    dispatch(pushToStack(path))
    dispatch(setCurrentDir(file._id))
  }

  

  const toggleDropdown = () => {
    setIsOpen(!isOpen)
  }

  return (
    <div className='diritem'>
      <div className="diritem__content">
        <div className="diritem__title"  onClick={() => openHandler(file)}>
          {title}
        </div>

        {hasChildren && (
          <div className="diritem__dropdown" onClick={toggleDropdown}>
            <DropdownButton />
          </div>
        )}
      </div>

      {hasChildren && (
        <div className={`diritem__subitem ${isOpen ? 'visible' : ''}`}>
          {file.childs.map((dir) =>
            <DirItem file={dir} title={dir.name} key={dir.path}/>
          )}
        </div>
      )}

    </div>
  )
}

export default DirItem