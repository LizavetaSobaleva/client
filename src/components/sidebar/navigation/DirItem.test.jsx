import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import DirItem from './DirItem';
import { pushToStack, setCurrentDir, clearStack } from '../../../reducers/fileReducer';

const mockStore = configureStore([]);
let store;

const file = {
  _id: '1',
  name: 'Root',
  idPath: ['1'],
  childs: [
    {
      _id: '2',
      name: 'Child 1',
      idPath: ['1', '2'],
      childs: []
    },
    {
      _id: '3',
      name: 'Child 2',
      idPath: ['1', '3'],
      childs: []
    }
  ]
};

const renderComponent = (store) =>
  render(
    <Provider store={store}>
      <DirItem file={file} title={file.name} />
    </Provider>
  );

describe('DirItem Component', () => {
  beforeEach(() => {
    store = mockStore({
      files: { currentDir: 'root' }
    });
    store.dispatch = jest.fn();
  });

  test('renders DirItem without crashing', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('dirItem-1')).toBeInTheDocument();
    expect(screen.getByTestId('dirItemTitle-1')).toHaveTextContent('Root');
    expect(asFragment()).toMatchSnapshot();
  });

  test('calls openHandler when title is clicked', () => {
    renderComponent(store);
    fireEvent.click(screen.getByTestId('dirItemTitle-1'));
    expect(store.dispatch).toHaveBeenCalledWith(clearStack());
    expect(store.dispatch).toHaveBeenCalledWith(pushToStack([]));
    expect(store.dispatch).toHaveBeenCalledWith(setCurrentDir('1'));
  });

  test('renders dropdown button if there are children', () => {
    const { asFragment } = renderComponent(store);
    expect(screen.getByTestId('dirItemDropdown-1')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('toggles subitems visibility when dropdown button is clicked', () => {
    renderComponent(store);
    const subItem = screen.getByTestId('dirItemSubItem-1');
    expect(subItem).not.toHaveClass('visible');
    fireEvent.click(screen.getByTestId('dirItemDropdown-1'));
    expect(subItem).toHaveClass('visible');
  });
});
