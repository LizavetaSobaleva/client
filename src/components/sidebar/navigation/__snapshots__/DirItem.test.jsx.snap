// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DirItem Component renders DirItem without crashing 1`] = `
<DocumentFragment>
  <div
    class="diritem"
    data-testid="dirItem-1"
  >
    <div
      class="diritem__content"
    >
      <div
        class="diritem__title"
        data-testid="dirItemTitle-1"
      >
        Root
      </div>
      <div
        class="diritem__dropdown"
        data-testid="dirItemDropdown-1"
      >
        <div
          class="dropdown"
          data-testid="dropdownButton"
        >
          <img
            alt=""
            class="dropdown__icon "
            data-testid="dropdownIcon"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
    <div
      class="diritem__subitem "
      data-testid="dirItemSubItem-1"
    >
      <div
        class="diritem"
        data-testid="dirItem-2"
      >
        <div
          class="diritem__content"
        >
          <div
            class="diritem__title"
            data-testid="dirItemTitle-2"
          >
            Child 1
          </div>
        </div>
      </div>
      <div
        class="diritem"
        data-testid="dirItem-3"
      >
        <div
          class="diritem__content"
        >
          <div
            class="diritem__title"
            data-testid="dirItemTitle-3"
          >
            Child 2
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`DirItem Component renders dropdown button if there are children 1`] = `
<DocumentFragment>
  <div
    class="diritem"
    data-testid="dirItem-1"
  >
    <div
      class="diritem__content"
    >
      <div
        class="diritem__title"
        data-testid="dirItemTitle-1"
      >
        Root
      </div>
      <div
        class="diritem__dropdown"
        data-testid="dirItemDropdown-1"
      >
        <div
          class="dropdown"
          data-testid="dropdownButton"
        >
          <img
            alt=""
            class="dropdown__icon "
            data-testid="dropdownIcon"
            src="test-file-stub"
          />
        </div>
      </div>
    </div>
    <div
      class="diritem__subitem "
      data-testid="dirItemSubItem-1"
    >
      <div
        class="diritem"
        data-testid="dirItem-2"
      >
        <div
          class="diritem__content"
        >
          <div
            class="diritem__title"
            data-testid="dirItemTitle-2"
          >
            Child 1
          </div>
        </div>
      </div>
      <div
        class="diritem"
        data-testid="dirItem-3"
      >
        <div
          class="diritem__content"
        >
          <div
            class="diritem__title"
            data-testid="dirItemTitle-3"
          >
            Child 2
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
